36-vending-machine-monitor/firmware/vending-machine/vending-machine.ino

```diff
@@ -0,0 +1,1125 @@
+            // empties, the distance to the first can increases.
+            if (d <= columnSize.fullDistance+halfItem) {
+                fillState = DispensingItems::COLUMN_FULL;
+            }
+            else if (d >= columnSize.emptyDistance-halfItem) {
+                fillState = DispensingItems::COLUMN_EMPTY;
+            }
+            else if (d >= columnSize.lowDistance+halfItem) {
+                fillState = DispensingItems::COLUMN_LOW;
+            }
+            else {
+                fillState = DispensingItems::COLUMN_GOOD;
+            }
+            if (columnSize.canHeight) {
+                if (columnSize.emptyDistance > distance.distance_mm) {
+                    // add half a can so that the nominal point is in the middle of the can
+                    // This prevents jitter from creating false readings.
+                    itemCount = (columnSize.emptyDistance-distance.distance_mm+halfItem) / columnSize.canHeight;
+                }
+                else {
+                    itemCount = 0; // womp womp, no more soda in this column
+                }
+            }
+            else {
+                itemCount = ITEM_COUNT_UNKNOWN;
+            }
+        }
+
+        items.itemCount = itemCount;
+        items.fillState = fillState;
+
+        if (items.itemCount != prevItems.itemCount || items.fillState != prevItems.fillState) {
+            columnStateChanged(prevItems, items);
+        }
+    }
+
+    /**
+     * @brief Adds fields to the body of an event for the state of this column.
+     * 
+     * @param items The state of the dispensing items
+     * @return J* The created event body
+     */
+    J* buildColumnEvent(const DispensingItems& items) {
+        J* body = JCreateObject();
+        JAddNumberToObject(body, DATA_FIELD_COL, columnNumber());
+        JAddStringToObject(body, DATA_FIELD_NAME, columnName());
+        JAddStringToObject(body, DATA_FIELD_STATE, items.fillStateString());
+        if (items.isOnline()) {
+            JAddNumberToObject(body, DATA_FIELD_DISTANCE, this->distance.distance_mm);
+            JAddNumberToObject(body, DATA_FIELD_ITEMS, items.itemCount);
+        }
+        return body;
+    }
+
+    /**
+     * @brief Creates and sends and event about the current column state, and what changed
+     *  since the last reported event.
+     * 
+     * @param prev      The previous column state
+     * @param current   The current column state
+     */
+    void columnStateChanged(const DispensingItems& prev, const DispensingItems& current) {
+        char changed[128];
+        *changed = 0;
+        J* body = buildColumnEvent(current);
+        if (current.isOnline() && prev.itemCount != current.itemCount) {
+            strncat(changed, ",items", sizeof(changed)-1);
+        }
+        if (prev.fillState != current.fillState) {
+            strncat(changed, ",state", sizeof(changed)-1);
```

Comment:
It doesn't matter in this case, because your buffer is big enough, but this second call to `strncat` should technically take into account the bytes that were potentially already added to the buffer at line 461. There's no longer `sizeof(changed)-1` bytes available to write to in that case. There's `sizeof(changed)-1-strlen(changed)` left.